src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2016, Shogun-Toolbox e.V. <shogun-team@shogun-toolbox.org>
+ * All rights reserved.
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Heiko Strathmann, Viktor Gal
+ */
+
+#include <shogun/mathematics/linalg/LinalgBackendBase.h>
+#include <shogun/mathematics/linalg/SGLinalg.h>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+/**
+ * Wrapper method for internal implementation of vector dot-product that works
+ * with generic vectors.
+ *
+ * @param a first vector
+ * @param b second vector
+ * @return the dot product of \f$\mathbf{a}\f$ and \f$\mathbf{b}\f$, represented
+ * as \f$\sum_i a_i b_i\f$
+ */
+template <typename Type>
+Type dot(const SGVector<Type>& a, const SGVector<Type>& b)
+{
+	if (a.on_gpu() && b.on_gpu())
+	{
+		return sg_linalg->get_gpu_backend()->dot(a, b);
+	}
+	else if (a.on_gpu() || b.on_gpu())
+		SG_SERROR("Cannot operate with one vector on CPU while the other one is on GPU.\n");
```

Comment:
`SG_SERROR("Cannot operate with first vector gpu (%d) and second vector gpu (%d).\n", a.on_gpu(), b.on_gpu());`